### PR TITLE
formula: add preset `pour_bottle?` symbols

### PR DESCRIPTION
**Pull request by Rylan12**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

See https://github.com/Homebrew/brew/issues/11302

This PR adds two symbols that can be passed to `pour_bottle?` as shortcuts for commonly used `pour_bottle?` blocks:

- `:default_prefix_required`: for when the bottle only works in the default prefix but relocation detection doesn't find any problematic paths (right now this would apply only to `ocaml`). Using this symbol is a shortcut for:
    ```ruby
    pour_bottle? do
      reason <<~EOS
        The bottle needs to be installed into #{Homebrew::DEFAULT_PREFIX}.
      EOS
      satisfy { HOMEBREW_PREFIX.to_s == Homebrew::DEFAULT_PREFIX }
    end
    ```
- `:clt_required`: for when the CLT is needed for bottles (right now this would be used by the `gcc` and `python` formulae). Using this symbol is a shortcut for:
    ```ruby
    pour_bottle? do
      reason <<~EOS
        The bottle needs the Apple Command Line Tools to be installed.
          You can install them, if desired, with:
            xcode-select --install
      EOS
      satisfy { HOMEBREW_PREFIX.to_s == Homebrew::DEFAULT_PREFIX }
    end
    ```

I'm venturing a bit into parts of Ruby that I'm less familiar with so its possible that there's a better way for this to be done that what I have here (e.g. maybe the options should go in the `PourBottleCheck` class rather than `Formula`). I've also needed to add some `T.cast`s which makes me wonder if there is a better way to accomplish this.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
> * `:default_prefix_required`: for when the bottle only works in the default prefix but relocation detection doesn't find any problematic paths (right now this would apply only to `ocaml`). Using this symbol is a shortcut for:

I wonder if this is worth adding for a single formula (particularly if we suspect we may be able to eliminate this).

> * `:clt_required`: for when the CLT is needed for bottles (right now this would be used by the `gcc` and `python` formulae). Using this symbol is a shortcut for:

Is it only these two? 👍🏻 on the name here.

File: Library/Homebrew/formula.rb [hunk 1]
```diff
@@ -2918,8 +2918,34 @@ def service(&block)
     #
     # If `satisfy` returns `false` then a bottle will not be used and instead
     # the {Formula} will be built from source and `reason` will be printed.
-    def pour_bottle?(&block)
+    #
+    # Alternatively, a preset reason can be passed as a symbol:
+    # <pre>pour_bottle? :default_prefix_required</pre>
+    # <pre>pour_bottle? :clt_required</pre>
+    def pour_bottle?(requirement = nil, &block)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:2925]:
```suggestion
    def pour_bottle?(reason: nil, &block)
```

maybe? We use that elsewhere and it feels appropriate here.

File: Library/Homebrew/formula.rb [hunk 2]
```diff
@@ -2918,8 +2918,34 @@ def service(&block)
...
       @pour_bottle_check = PourBottleCheck.new(self)
+
+      block ||= case requirement
+      when :default_prefix_required
+        lambda do |_|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:2930]:
```suggestion
        lambda do
```

Does this work?

**Comment by Rylan12** [comment]
> I wonder if this is worth adding for a single formula (particularly if we suspect we may be able to eliminate this).

Yeah, it probably isn't. I was thinking that it would be worth adding because it's a more common reason but, yeah, most (if not all, soon) have been removed. It's probably better to just continue to improve the logic that determines if a bottle's relocatable to catch more of these situations, anyway.

> Is it only these two? 👍🏻 on the name here.

Yep, this should cover all but one `pour_bottle?` block in homebrew/core. The remaining block is in [`percona-server`](https://github.com/Homebrew/homebrew-core/blob/215a1061d9d98c8b31ec14b2eee9e9477901a720/Formula/percona-server.rb#L20-L23). I guess we could add a reason for this but it's the same issue as with `:default_prefix_required` (plus it feels like a very specific reason).

File: Library/Homebrew/formula.rb (hunk 2 above)
Comment (Rylan12) [reply Library/Homebrew/formula.rb:2930]: No, I get this error:

```
FormulaUnreadableError (hello: wrong number of arguments (given 1, expected 0))
```

According to the [`instance_eval` docs](https://ruby-doc.org/core-2.5.1/BasicObject.html#method-i-instance_eval):

> When `instance_eval` is given a block, _obj_ is also passed in as the block's only argument.

**Comment by Rylan12** [comment]
Two more thoughts, if a user supplies both a reason and a block, what should happen? Right now the reason will take priority. I wonder, though, if it should be the other way around (or maybe just an error).

Also, should this throw an error if an invalid reason is specified?

**Review by MikeMcQuaid — DISMISSED** [review]

**Comment by MikeMcQuaid** [comment]
> if it should be the other way around (or maybe just an error).

Yes!

> Also, should this throw an error if an invalid reason is specified?

Yes!

> Yep, this should cover all but one `pour_bottle?` block in homebrew/core. The remaining block is in [`percona-server`](https://github.com/Homebrew/homebrew-core/blob/215a1061d9d98c8b31ec14b2eee9e9477901a720/Formula/percona-server.rb#L20-L23). I guess we could add a reason for this but it's the same issue as with `:default_prefix_required` (plus it feels like a very specific reason).

Agreed 👍🏻

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by Rylan12** [comment]
Once there's been a new tag I'll start making these changes in homebrew/core

**Pull request merged**